gpbft/chain.go

```diff
@@ -5,16 +5,76 @@ import (
 	"encoding/binary"
 	"encoding/hex"
 	"errors"
+	"fmt"
 	"strings"
+
+	cbg "github.com/whyrusleeping/cbor-gen"
+	"golang.org/x/crypto/blake2b"
 )
 
+// TipSetKey is a
+type TipSetKey = []byte
+
+type CID = []byte
+
+var cidPrefix = []byte{0x01, 0x71, 0xA0, 0xE4, 0x02, 0x20}
+
+// Hashes the given data and returns a CBOR + blake2b-256 CID.
+func MakeCid(data []byte) []byte {
+	// TODO: Consider just using go-cid? We implicitly depend on it through cbor-gen anyways.
+	digest := blake2b.Sum256(data)
+
+	out := make([]byte, 0, 38)
+	out = append(out, cidPrefix...)
+	out = append(out, digest[:]...)
+	return out
+}
+
 // Opaque type representing a tipset.
 // This is expected to be:
 // - a canonical sequence of CIDs of block headers identifying a tipset,
 // - a commitment to the resulting power table,
 // - a commitment to additional derived values.
 // However, GossipPBFT doesn't need to know anything about that structure.
-type TipSet = []byte
+type TipSet struct {
+	Epoch       int64
+	TipSet      TipSetKey
+	PowerTable  CID
+	Commitments [32]byte
+}
+
+func (ts *TipSet) IsZero() bool {
+	return len(ts.TipSet) == 0
+}
+
+func (ts *TipSet) Equal(b *TipSet) bool {
+	return ts.Epoch == b.Epoch &&
+		bytes.Equal(ts.TipSet, b.TipSet) &&
+		bytes.Equal(ts.PowerTable, b.PowerTable) &&
+		ts.Commitments == b.Commitments
+}
+
+func (ts *TipSet) MarshalForSigning() []byte {
+	var buf bytes.Buffer
+	_ = cbg.WriteByteArray(&buf, ts.TipSet)
```

Comment:
Why is this using `cbg`? Just `buf.Write`?

Reply:
Because https://github.com/filecoin-project/FIPs/blob/b98ae5f0e9c0b033279f29fa5c507e457f826947/FIPS/fip-0086.md?plain=1#L307

I'm actually not sure where the wires got crossed in the FEVM FIP. IMO, it should have either been `CBOR([Cid, Cid, Cid])` or `Raw(Cid || Cid || Cid)`. Instead we got `CBOR(Cid || Cid || Cid)`.

Reply:
Honestly, I'm tempted to change it to `CBOR([Cid, Cid, Cid])` because then we'd be able to dereference the CIDs. E.g.,`$TIPSET_CID/0` would refer to the CID of the first block. But if we did that, we'd probably (eventually) want a small FIP to make FEVM use the new format.